### PR TITLE
Changes to make Megatron NMT exportable

### PR DESCRIPTION
**Pull request by Davood-M**
# What does this PR do ?

Adding requirements to make Megatron NMT exportable

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Added input/output types for encoder/decoder/tokens_head
- Added input_example

# Before your PR is "Ready for review"
**Pre checks**:
- [X] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [] New Feature
- [X] Bugfix
- [ ] Documentation

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

### CODE REVIEW

File: nemo/collections/nlp/modules/common/megatron/fused_bias_gelu.py [hunk 1]
```diff
@@ -62,6 +62,10 @@ def backward(ctx, grad_output):
         tmp = bias_gelu_back(grad_output, bias, input)
         return tmp, tmp
 
+    @staticmethod
```
Comment (borisfom) [inline nemo/collections/nlp/modules/common/megatron/fused_bias_gelu.py:65]: Does that work for ONNX->TRT translation or only for ORT ?

**Review by MaximumEntropy — COMMENTED** [review]
Minor comment on dtype setting.

**Comment by MaximumEntropy** [comment]
@ericharper what do you think about having separate `xx_module.py` files for the different modules versus moving all of them into a single file?

**Review by MaximumEntropy — CHANGES_REQUESTED** [review]
Just marking as needs changes temporarily to get @ericharper's review.

File: nemo/utils/export_utils.py [hunk 2]
```diff
@@ -168,7 +168,10 @@ def replace_FusedLayerNorm(n: nn.Module) -> Optional[nn.BatchNorm2d]:
             return None
 
         dev = next(n.parameters()).device
-        mod = nn.LayerNorm(n.normalized_shape, eps=n.eps, elementwise_affine=n.elementwise_affine,).to(dev)
+        if isinstance(n, FusedLayerNorm) or isinstance(n, MixedFusedLayerNorm):
+            mod = nn.LayerNorm(n.normalized_shape, eps=n.eps, elementwise_affine=n.elementwise_affine,).to(dev)
+        elif isinstance(n, FastLayerNorm):
+            mod = nn.LayerNorm(n.weight.shape, eps=n.epsilon, elementwise_affine=True, dtype=torch.float16,).to(dev)
```
[the remark this answers is not in the document]
Comment (Davood-M) [reply nemo/utils/export_utils.py:202]: Yes. Without setting the dtype, it raises the following error:

```python
Traceback (most recent call last):
  File "***", line 245, in <module>
    encoder_instance.export('***/encoder.onnx')
  File "***/NeMo/nemo/core/classes/exportable.py", line 68, in export
    out, descr, out_example = model._export(
  File "***/NeMo/nemo/core/classes/exportable.py", line 142, in _export
    output_example = tuple(self.forward(*input_list, **input_dict))
  File "nemo2onnx-experimental.py", line 166, in forward
    return self.encoder(
  File "***/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "***/NeMo/nemo/collections/nlp/modules/common/megatron/megatron_transformer_encoder.py", line 156, in forward
    enc_output = self.model(
  File "***/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "***/NeMo/nemo/collections/nlp/modules/common/megatron/transformer.py", line 1819, in forward
    output = self.final_layernorm(output)
  File "***/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "***/NeMo/nemo/utils/export_utils.py", line 56, in forward
    ret = self.mod.forward(x)
  File "***/lib/python3.8/site-packages/torch/nn/modules/normalization.py", line 189, in forward
    return F.layer_norm(
  File "***/lib/python3.8/site-packages/torch/nn/functional.py", line 2486, in layer_norm
    return torch.layer_norm(input, normalized_shape, weight, bias, eps, torch.backends.cudnn.enabled)
RuntimeError: expected scalar type Half but found Float
```

File: nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py [hunk 3]
```diff
@@ -52,6 +55,179 @@
...
+    def output_names(self) -> List[str]:
+        return ['log_probs']
+
+
+class DecWEmb(torch.nn.Module, Exportable):
+    """
+    Combines decoder_embedding with the decoder component
+    """
+
+    def __init__(self, decoder_embedding, decoder, device='cuda:0'):
```
Comment (ericharper) [inline nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py:114]: See device comment above.

File: nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py [hunk 4]
```diff
@@ -52,6 +55,179 @@
...
+    def output_names(self) -> List[str]:
+        return ['last_hidden_states']
+
+
+class EncWEmb(torch.nn.Module, Exportable):
+    """
+    Combines encoder_embedding with the encoder component
+    """
+
+    def __init__(self, encoder_embedding, encoder, device='cuda:0'):
```
Comment (ericharper) [inline nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py:181]: See device comment above.

File: nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py [hunk 5]
```diff
@@ -52,6 +55,179 @@
...
+    @property
+    def input_names(self) -> List[str]:
+        return ['hidden_states']
+
+    @property
+    def output_names(self) -> List[str]:
+        return ['log_probs']
+
+
+class DecWEmb(torch.nn.Module, Exportable):
```
Comment (ericharper) [inline nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py:109]: See module comment above.

File: nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py [hunk 6]
```diff
@@ -52,6 +55,179 @@
...
+    @property
+    def input_names(self) -> List[str]:
+        return ['input_ids', 'decoder_mask', 'encoder_mask', 'encoder_embeddings', 'decoder_memes']
+
+    @property
+    def output_names(self) -> List[str]:
+        return ['last_hidden_states']
+
+
+class EncWEmb(torch.nn.Module, Exportable):
```
Comment (ericharper) [inline nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py:176]: See module comment above.

File: nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py [hunk 7]
```diff
@@ -801,3 +977,16 @@ def on_validation_start(self) -> None:
...
+    def encoder(self):
+        return EncWEmb(self.enc_dec_model.encoder_embedding, self.enc_dec_model.enc_dec_model.encoder)
+
+    @property
+    def decoder(self):
+        return DecWEmb(self.enc_dec_model.decoder_embedding, self.enc_dec_model.enc_dec_model.decoder)
+
+    @property
+    def tokens_head(self):
+        return TokensHeadWEmb(self.enc_dec_model.decoder_embedding, self.enc_dec_model.tokens_head)
```
Comment (ericharper) [inline nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py:992]: Shouldn't these be going into a base class? Also, can you consider changing the suffix `WEmb` Maybe just remove the `W`?

File: nemo/collections/nlp/modules/common/megatron/megatron_decoder_module.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            "enc_attn_mask": NeuralType(('B', 'T'), MaskType()),
+        }
+
+    @property
+    def input_names(self) -> List[str]:
+        return ['dec_input', 'dec_attn_mask', 'enc_output', 'enc_attn_mask']
+
+    @property
+    def output_names(self) -> List[str]:
+        return ['decoder_output']
```
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/megatron_decoder_module.py:43]: Why do you need names property? Don't you already have the name in the input_type?

File: nemo/collections/nlp/modules/common/megatron/megatron_encoder_module.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+            "encoder_mask": NeuralType(('B', 'T'), MaskType()),
+        }
+
+    @property
+    def input_names(self) -> List[str]:
+        return ['input_ids', 'encoder_mask']
+
+    @property
+    def output_names(self) -> List[str]:
+        return ['encoder_output']
```
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/megatron_encoder_module.py:43]: See names comment above.

File: nemo/collections/nlp/modules/common/megatron/megatron_tokens_head_module.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+            "embeddings_weights": NeuralType(('T', 'D'), MaskType()),
+        }
+
+    @property
+    def input_names(self) -> List[str]:
+        return ['dec_output', 'embeddings_weights']
+
+    @property
+    def output_names(self) -> List[str]:
+        return ['logits']
```
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/megatron_tokens_head_module.py:41]: See name comment above.

**Review by ericharper — CHANGES_REQUESTED** [review]
Finished my first review. Overall a great PR! But I think it still need some refactoring.

Please see my comments.

**Comment by Davood-M** [comment]
reopening PR

**Pull request merged**